crypto/stark-verifier-ethereum/contracts/fri.sol

```diff
@@ -29,20 +83,203 @@ contract Fri {
                 }
             }
         }
-        sort(queries);
+        queries.sort();
         return queries;
     }
 
-    // This function sorts the array
-    // Note - We use insertion sort, the array is expected to be small so this shouldn't
-    // cause problems.
-    function sort(uint64[] memory data) internal pure {
-        for (uint256 i = 0; i < data.length; i++) {
-            uint256 j = i;
-            while (j > 0 && data[j] < data[j - 1]) {
-                (data[j], data[j - 1]) = (data[j - 1], data[j]);
-                j--;
+    // Unwraping endpoint because the main function has too deep of a stack otherwise
+    function fri_check(
+        ProofTypes.StarkProof memory proof,
+        uint8[] memory fri_layout,
+        uint256[] memory eval_points,
+        uint8 log_eval_domain_size,
+        uint64[] memory queries,
+        uint256[] memory polynomial_at_queries,
+        uint256 oods_point,
+        uint256 evaluated_oods_point
+    ) internal {
+        fold_and_check_fri_layers(
+            FriContext(
+                proof.fri_values,
+                proof.fri_commitments,
+                proof.fri_decommitments,
+                fri_layout,
+                eval_points,
+                log_eval_domain_size,
+                queries,
+                polynomial_at_queries,
+                proof.last_layer_coeffiencts
+            )
+        );
+
+        // The final check is that the constraints evaluated at the out of domain sample are
+        // equal to the values commited constraint values
+        uint256 result = 0;
+        uint256 power  = 0x0000000000000000000000000000000000000000000000000000000000000001.to_montgomery();
+        for (uint256 i = 0; i < proof.constraint_oods_values.length; i++) {
+            result = result.fadd(proof.constraint_oods_values[i].fmul_mont(power));
+            power = power.fmul_mont(oods_point);
+        }
+        require(result == evaluated_oods_point, "Oods mismatch");
+    }
+
+    // This function takes in fri values, decommitments, and layout and checks the folding and merkle proofs
+    // Note the final layer folded values will live in the
+    function fold_and_check_fri_layers(FriContext memory fri_data) internal {
+        Eval_X memory eval = init_eval(fri_data.log_eval_domain_size);
+        LayerContext memory layer_context = LayerContext({
+            len: uint64(2)**(fri_data.log_eval_domain_size),
+            step: 1,
+            coset_size: 0
+        });
+        uint256[] memory merkle_ind = new uint256[](fri_data.queries.length);
```

Comment:
`idx` is the 'standard' three letter abbreviation of 'index'. Even better is to just use `merkle_index` and `merkle_value`.